Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,47 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+# True if this version of Ruby supports text encodings in its strings
+def ruby_has_encoding?
+  String.method_defined?(:force_encoding)
+end
+
+# Truncates a text string to fit within a byte size constraint,
+# preserving character encoding validity. The returned string will
+# be not much longer than the specified max_bytes, though the exact
+# shortfall or overrun may vary.
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options.fetch(:front_weight, 0.5)
+  if front_weight < 0.0 || front_weight > 1.0
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  return s if s.bytesize <= max_bytes
+
+  glue = "\n[...snip...]\n"
+  max_bytes_in = [max_bytes - glue.bytesize, 1].max
+  if ruby_has_encoding?
+    bytes = s.dup.force_encoding("BINARY")
+    glue_bytes = glue.encode("BINARY")
+  else
+    bytes = s
+    glue_bytes = glue
+  end
+  if front_weight == 0.0
+    front = bytes[1..0]
+    back = bytes[-max_bytes_in..-1]
+  elsif front_weight == 1.0
+    front = bytes[0..max_bytes_in]
+    back = bytes[1..0]
+  else
+    front = bytes[0..(max_bytes_in * front_weight).floor]
+    back = bytes[-(max_bytes_in * (1.0 - front_weight)).ceil..-1]
+  end
```

Comment:
You might want to modify this a bit to avoid other (less likely) edge cases, e.g. if `front_weight != 0.0`, but `(max_bytes_in * front_weight).floor == 0`:

```ruby
front_bytes = (max_bytes_in * front_weight).floor
back_bytes = max_bytes_in - front_bytes
if front_bytes == 0
  front = bytes[1..0]
  back = bytes[-max_bytes_in..-1]
elsif back_bytes == 0
  front = bytes[0..(max_bytes_in - 1)]
  back = bytes[1..0]
else
  front = bytes[0..(front_bytes - 1)]
  back = bytes[-back_bytes..-1]
end
```

Notice also the slight change in the upper limit for `front` (added `- 1`). Just a very minor detail, but otherwise you'll get one more byte than desired. (We're in the popular category of off-by-one errors. 😄)